plugins/action/iptables_state.py

```diff
@@ -0,0 +1,164 @@
+# Copyright: (c) 2020, quidame <quidame@poivron.org>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+import time
+
+from ansible.plugins.action import ActionBase
+from ansible.errors import AnsibleError, AnsibleActionFail, AnsibleConnectionFailure
+from ansible.utils.vars import merge_hash
+from ansible.utils.display import Display
+
+display = Display()
+
+
+class ActionModule(ActionBase):
+
+    # Keep internal params away from user interactions
+    _VALID_ARGS = frozenset(('path', 'state', 'table', 'noflush', 'counters', 'modprobe', 'ip_version', 'wait'))
+    DEFAULT_SUDOABLE = True
+
+    MSG_ERROR__ASYNC_AND_POLL_NOT_ZERO = (
+        "This module doesn't support async>0 and poll>0 when its 'state' param "
+        "is set to 'restored'. To enable its rollback feature (that needs the "
+        "module to run asynchronously on the remote), please set task attribute "
+        "'poll' to 0, and 'async' to a value not greater than 'ansible_timeout' "
+        "(recommended).")
+    MSG_WARNING__NO_ASYNC_IS_NO_ROLLBACK = (
+        "Attempts to restore iptables state without rollback in case of mistake "
+        "may lead the ansible controller to loose access to the hosts and never "
+        "regain it before fixing firewall rules through a serial console, or any "
+        "other way except SSH. Please set task attribute 'poll' to 0, and 'async' "
+        "to a value not greater than 'ansible_timeout' (recommended).")
+
+    def _async_result(self, module_args, task_vars, timeout):
+        '''
+        Retrieve results of the asynchonous task, and display them in place of
+        the async wrapper results (those with the ansible_job_id key).
+        '''
+        for i in range(timeout):
+            async_result = self._execute_module(
+                module_name='async_status',
+                module_args=module_args,
+                task_vars=task_vars,
+                wrap_async=False)
+            if async_result['finished'] == 1:
+                break
+            time.sleep(1)
+
+        del async_result['ansible_job_id']
+        del async_result['finished']
+
+        return async_result
+
+    def run(self, tmp=None, task_vars=None):
+
+        # individual modules might disagree but as the generic the action plugin, pass at this point.
+        self._supports_check_mode = True
+        self._supports_async = True
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        del tmp  # tmp no longer has any effect
+
+        task_async = self._task.async_val
+        task_poll = self._task.poll
+        module_name = self._task.action
+        module_args = self._task.args
+
+        if not result.get('skipped'):
+
+            if result.get('invocation', {}).get('module_args'):
```

Comment:
Where do you handle the case that the module invocation failed (for example because invalid module options were passed)?

Reply:
valid/invalid params are filtered by frosenset() in action plugin, and their values are handled by the module itself. I added some basic checks in tests right now.

Reply:
I mean invalid **values** for options. These are only handled by the module itself. It doesn't make sense to re-implement validation on the controller.

Reply:
I do not really validate options in the plugin, but instead task attributes (check_mode, async, poll) and their consistency with some options. Conflict between `check_mode` and `async` is managed elsewhere by ansible, but there are some other cases, as the module may need to run asynchronously when `sate: restored`. But the action plugin doesn't check what is a valid value/type for the module options. It just sets internal params to put in the module call.

Reply:
Is there something wrong with this ?

Reply:
As I said: you need to react on the case that the module (which does more strict validation and fails if something is not right) fails. I still don't see that you check whether the module failed or succeeded.

Reply:
Hi, maybe I don't understand and need more clarification.
At first, I started by copying the `normal` action plugin (not a good idea ?) that does the same: if task is not skipped, then run module, and retrieve success or failure status from the module's results.
I just added the needed stuff to put internal parameters to the module before calling it, and to retrieve the results (and even the async results), filter them, and then finish the same way than the `normal` module.
Why should the plugin specifically differentiate failure or success if the module does this on its own ? Currently, the action plugin just 'pipes' the module results to the user, and I don't understand what is missing. Basically, if the module fails... the tasks fails. What could the plugin do more with that (or against that ?). Do you mean there is something to do with ''if result['failed']: ..." and/or with "if not result['failed']: ..." ?

Thanks for your patience

Reply:
Could you elaborate, please ? I still don't understand what you expect.
On the controller:
```
get a first, temporary result about valid task options and attributes
if the task is not skipped:
    if module's param "state" has value "restored":
        if task attributes "async" and "poll" allow this and are consistent with the value of "ansible_timeout" (otherwise fail or display a warning):
            set module's param "_timeout" to the same value than "async" task attribute
            set module's param "_back" to a value relative to "async_dir"

    execute the module and retrieve its result (that may be, in our case, a temporary result)

    if module's param "_back" is defined (that means the module is executed asynchrounously):
        reset the connection
        in a loop of "ansible_timeout" iterations of 1s:
            try to remove "_back" on the host
            (during this time, on the host, check every 1s if "_back" exists, or return as succeeded. After "_timeout" iterations, if "_back" still exist, restore the iptables state from this backup, and return as failed)
        retrieve the async result of the module (and merge it with the previous, temporary result)
        remove async wrapper's specific objects from the result
        remove internal param ("_back" and "_timeout") from the result

return result
```
So, what is wrong with this, and where ?
1. Is there something to remove from the plugin, because it is already (or should) be implemented in the module ?
2. Is there something to add in the plugin, because... just... it is expected to be there ? And so, what is missing ?

> you need to react on the case that the module (which does more strict validation and fails if something is not right) fails. I still don't see that you check whether the module failed or succeeded.

Does it mean I have to parse the last result again, and override it in case of failure or success ? What's the benefit ?
Does it mean, instead, that I have to catch errors about the module invocation itself, in some try/except construct ? (I hopped that the `self._execute_module()` did that on its own ! maybe I'm wrong)
Does it mean I have to open other conditionals at the same level of "if the task is not skipped", to display some meaningful messages in case of unsupported attributes or options ? (I hopped that `super(ActionModule, self).run(tmp, task_vars)` did that on its own ! maybe I'm wrong)
Does it mean something else ? Could you point me to some useful example, please ?
Thanks

Reply:
As I said: you need to react to the case that executing the module fails. You do not seem to be doing that. The `normal` plugin doesn't need to do this since it just passes the information on, it doesn't do post-processing which requires that the execution succeeded.

It's the line `if module's param "_back" is defined (that means the module is executed asynchrounously):` that needs to get an `and invocation of module did not fail`.

Reply:
Hi @felixfontein , ok, I understand what you mean, now. But I still don't think so.

What we are talking about is the rollback to previous rules... To allow the module to restore the initial state of the firewall in case of bad ruleset (the controller is unable to access the host after connection reset is the criteria to say "bad ruleset"), the module keeps a backup of this initial state. The other part is to try to remove this backup from the plugin (that is a proof, in itself, that the controller can access the host). To achieve that, not only the module must be executed asynchronously, but also its result *must not be polled* by the async wrapper itself. If the result was polled by async wrapper, what you call post-processing would be processed *after* polling, that means **too late**: after the rollback happened on the host, even with any "good ruleset".

So, `poll=0` is required (if not the case, the plugin raises an error). And if `poll=0`, `self._execute_module(...)` will **never** report the actual result of the module (even early failure due to either a missing required option, a bad option value/type, etc) but only, and always, the async wrapper result, crude (ansible_job_id, results_file, started=1, finished=0), and no more. This is what is needed to retrieve the module result after processing the 2 loops, **one on each side, at the same time**. The rollback is a kind of race between the plugin and the module, rather than two steps. The plugin has to do something, and it has to do it without worrying about the module result, because the module result depends on the plugin's *action* (that is to remove the iptables backup on the host before the module restores it and then returns as failed).

So, yes, a part of the plugin's code (reset connection, remove the backup on the host) is executed blindly, and for me it makes sense: it is executed even if the host becomes unreachable. Anyway, if the module fails early (i.e. before attempting to restore iptables state), this failure will be returned as well... (unfortunately with a time offset of 1s = 1 loop iteration).

Does it make sense for you ?

Reply:
I finally check (once) whether or not the asynchronous task is already *finished* (success or failure don't really matter) before resetting connection and trying in loop to remove the backup. I'm not sure it's a real benefit, but I'm not sure it is not, too.

Reply:
Ok, sounds reasonable.